miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java

```diff
@@ -506,4 +527,160 @@ public void setRunDao(RunStore runDao) {
     this.runDao = runDao;
   }
 
+  @Override
+  public boolean processNotification(NotificationDto dto) throws IOException, MisoNamingException {
+    final Date now = new Date();
+    User user = securityManager.getUserByLoginName("notification");
+    SecurityContextHolder.getContext().setAuthentication(new SuperuserAuthentication(user));
+    final Run run;
+
+    Run runFromDb = runDao.getByAlias(dto.getRunName());
+    boolean isNew;
+
+    if (runFromDb == null) {
+      run = dto.getPlatformType().createRun(user);
+      run.setCreationTime(now);
+      run.setCreator(user);
+      run.setName(generateTemporaryName());
+      run.setAlias(dto.getRunName());
+      isNew = true;
+    } else {
+      run = runFromDb;
+      if (dto.getPlatformType() != run.getPlatformType()) {
+        throw new IllegalStateException("Run scanner detected a run from " + dto.getPlatformType().name() + " and there is a save run from "
+            + run.getPlatformType().name());
+      }
+      isNew = false;
+    }
+
+    run.setLastModifier(user);
+    boolean isMutated = false;
+    isMutated |= updateField(convertDate(dto.getCompletionDate()), run.getCompletionDate(), run::setCompletionDate);
+    isMutated |= updateField(dto.getMetrics().toString(), run.getMetrics(), run::setMetrics);
+    isMutated |= updateField(dto.getSequencerFolderPath().toString(), run.getFilePath(), run::setFilePath);
+    isMutated |= updateField(convertDate(dto.getStartDate()), run.getCompletionDate(), run::setCompletionDate);
+
+    final SequencerReference sequencer = sequencerReferenceService.getByName(dto.getSequencerName());
+    if (sequencer == null) {
+      throw new IllegalArgumentException("No such sequencer: " + dto.getSequencerName());
+    }
+    run.setSequencerReference(sequencer);
+
+    final Collection<SequencerPartitionContainer> containers = containerService.listByBarcode(dto.getContainerId());
+    switch (containers.size()) {
+    case 0:
+      SequencerPartitionContainer newContainer = new SequencerPartitionContainerImpl(user);
+      newContainer.setPlatform(sequencer.getPlatform());
+      newContainer.setCreator(user);
+      newContainer.setIdentificationBarcode(dto.getContainerId());
```

Comment:
Similarly, I think there's ambiguity with `containerId` -- I would have expected it would be a Long, and wouldn't have expected it to be the serial number...